protocol/communities/community_events_processing.go

```diff
@@ -0,0 +1,323 @@
+	if err != nil {
+		return err
+	}
+
+	e.filterInvalidEvents()
+	e.mergeEvents()
+	e.retainNewestEventsPerPropertyType()
+	e.sortEvents()
+	e.applyEvents()
+
+	return nil
+}
+
+func (e *eventsProcessor) validateDescription() error {
+	description, err := validateAndGetEventsMessageCommunityDescription(e.message.EventsBaseCommunityDescription, e.community.ControlNode())
+	if err != nil {
+		return err
+	}
+
+	if description.Clock != e.community.config.CommunityDescription.Clock {
+		return ErrInvalidCommunityEventClock
+	}
+
+	return nil
+}
+
+// Validate message's events, signature and details, filters out invalid ones.
+func (e *eventsProcessor) filterInvalidEvents() {
+	validateEvent := func(event *CommunityEvent) error {
+		signer, err := event.RecoverSigner()
+		if err != nil {
+			return err
+		}
+
+		err = e.community.ValidateEvent(event, signer)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	for i := range e.message.Events {
+		event := e.message.Events[i]
+
+		if err := validateEvent(&event); err == nil {
+			e.eventsToApply = append(e.eventsToApply, event)
+		} else {
+			e.logger.Warn("invalid community event", zap.Error(err))
+		}
+	}
+}
+
+// Merge message's events with community's events.
+func (e *eventsProcessor) mergeEvents() {
+	if e.community.config.EventsData != nil {
+		e.eventsToApply = append(e.eventsToApply, e.community.config.EventsData.Events...)
+	}
+}
+
+// Keep only the newest event per PropertyTypeID.
+func (e *eventsProcessor) retainNewestEventsPerPropertyType() {
+	eventsMap := make(map[string]CommunityEvent)
+
+	for _, event := range e.eventsToApply {
+		if existingEvent, found := eventsMap[event.PropertyTypeID()]; !found || event.CommunityEventClock > existingEvent.CommunityEventClock {
+			eventsMap[event.PropertyTypeID()] = event
+		}
+	}
+
+	e.eventsToApply = []CommunityEvent{}
+	for _, event := range eventsMap {
+		e.eventsToApply = append(e.eventsToApply, event)
+	}
+}
+
+// Sorts events by clock.
+func (e *eventsProcessor) sortEvents() {
+	sort.Slice(e.eventsToApply, func(i, j int) bool {
+		if e.eventsToApply[i].CommunityEventClock == e.eventsToApply[j].CommunityEventClock {
+			return e.eventsToApply[i].Type < e.eventsToApply[j].Type
+		}
+		return e.eventsToApply[i].CommunityEventClock < e.eventsToApply[j].CommunityEventClock
+	})
+}
+
+func (e *eventsProcessor) applyEvents() {
+	if e.community.config.EventsData == nil {
+		e.community.config.EventsData = &EventsData{
+			EventsBaseCommunityDescription: e.message.EventsBaseCommunityDescription,
+		}
+	}
+	e.community.config.EventsData.Events = e.eventsToApply
+
+	e.community.applyEvents()
+}
+
+func (o *Community) applyEvents() {
+	if o.config.EventsData == nil {
+		return
+	}
+
+	for _, event := range o.config.EventsData.Events {
+		err := o.applyEvent(event)
+		if err != nil {
+			o.config.Logger.Warn("failed to apply event", zap.Int32("type", int32(event.Type)), zap.Uint64("clock", event.CommunityEventClock))
```

Comment:
Should log the `err`?

Reply:
Failure when applying event is expected, that's more a warning to me.